core/src/main/java/org/infinispan/container/offheap/ExpandingOffHeapConcurrentMap.java

```diff
@@ -0,0 +1,1131 @@
+      // We only use the head pointer for the first iteration
+      long address = bucketHeadAddress;
+      InternalCacheEntry<WrappedBytes, WrappedBytes> ice = null;
+
+      while (address != 0) {
+         long nextAddress = offHeapEntryFactory.getNext(address);
+         // If the actualAddress was not known, check key equality otherwise just compare with the address
+         if (offHeapEntryFactory.equalsKey(address, key)) {
+            if (oldValue != null) {
+               ice = offHeapEntryFactory.fromMemory(address);
+               // If value doesn't match and was provided then don't replace it
+               if (!ice.getValue().equalsWrappedBytes(oldValue.getValue())) {
+                  ice = null;
+                  break;
+               }
+            }
+            // Need to always return the previous, so make sure we read it
+            if (ice == null) {
+               ice = offHeapEntryFactory.fromMemory(address);
+            }
+
+            long newAddress = offHeapEntryFactory.create(key, hashCode, newValue.getValue(), newValue.getMetadata());
+
+            entryReplaced(newAddress, address);
+            if (prevAddress != 0) {
+               offHeapEntryFactory.setNext(prevAddress, newAddress);
+            } else {
+               memoryLookup.putMemoryAddressOffset(memoryOffset, newAddress);
+            }
+            // We always set the next address on the newly created address - this will be 0 if the previous value
+            // was the end of the linked list
+            offHeapEntryFactory.setNext(newAddress, nextAddress);
+            break;
+         }
+         prevAddress = address;
+         address = nextAddress;
+      }
+      return ice;
+   }
+
+   @Override
+   public Set<WrappedBytes> keySet() {
+      throw new UnsupportedOperationException("keySet is not supported as it doesn't contain expiration data");
+   }
+
+   @Override
+   public Collection<InternalCacheEntry<WrappedBytes, WrappedBytes>> values() {
+      return new AbstractCollection<InternalCacheEntry<WrappedBytes, WrappedBytes>>() {
+         @Override
+         public Iterator<InternalCacheEntry<WrappedBytes, WrappedBytes>> iterator() {
+            return entryIterator();
+         }
+
+         @Override
+         public int size() {
+            return ExpandingOffHeapConcurrentMap.this.size();
+         }
+
+         @Override
+         public boolean remove(Object o) {
+            return o instanceof InternalCacheEntry && ExpandingOffHeapConcurrentMap.this.remove(((InternalCacheEntry) o).getKey(),
+                  ((InternalCacheEntry) o).getValue());
+         }
+      };
+   }
+
+   private class IterationPosition implements Iterator<InternalCacheEntry<WrappedBytes, WrappedBytes>> {
+      int bucketPosition;
+      int bucketCount = -1;
+      int bucketLockStop;
+      long bytesRead;
+
+      // This is arbitrary and can be changed
+      static final long READ_THRESHOLD = 8_000_000;
```

Comment:
> Maybe we do want it smaller for native, where there's less available RAM?

Yeah I am thinking of cutting it half. Hoping that is sufficient.

> Ideally for me we wouldn't need any buffering at all, because we'd read one bucket at a time and a bucket has <= 1 entry most of the time.

Not sure what you mean here. This is to allow us to read from multiple buckets in a single lock acquisition but stop if we are reading too many entries at once. We want to if possible read all buckets for a given lock region, but we don't want to if it reads too much memory.

Reply:
I'm thinking since we have 2x more locks than cores and no blocking with a lock held we should have little to no contention, so it should be ok to acquire the lock once per bucket (well, once for non-empty bucket). In order to read all of a lock region in one go we need to allocate all the entries at once and add them to the deque, and that might be more expensive than acquiring the lock repeatedly when there's no contention.

Reply:
We can't know a bucket is non-empty without holding a lock. Are you proposing we would lock the read lock, find the first memory bucket that is non zero and return the entries in that bucket and unlock?

And why do you say it is more expensive? If they don't need the full iteration? Or you are saying because the ArrayDeque may have to expand? Eventually all the entries would have to be read, with the PR currently the CPU should have the memory lookups to the buckets in cache for the entire "iteration". 

Also by not reading an entire lock region in one go, we open ourselves to a resize operation occurring in the middle of our iteration. Although that shouldn't happen very frequently.

Reply:
Exactly, acquiring the read lock for every non-empty bucket.

WRT performance I was thinking of the quarkus/native case, where we expect the heap memory to be limited and the GC to be a little slower.

Reply:
So running some JMH tests.

 entryAmount | diff 
----|----
1 | -5.154
100 | 9.427
2000 | 9.427
50000 | 3.135
250000 | 6.492
1000000 |  6.986

From the given run my guess as to why 1 was faster, because the latency to return a single entry would be quicker (but I don't know why for sure).

Reply:
However the more I think about it the loss of performance would make the code a bit cleaner and we would be putting less allocations at one time. I guess I can lean either way in regards to which is better. But for the same of limiting memory pressure I can make it do a single bucket at a time.

Reply:
Oh, actually, I forgot another reason I don't do this approach - the `expensiveIteration` is really really really expensive only doing 1 bucket at time. But this should happen very infrequently, so maybe that is okay?

Reply:
Actually `expensiveIteration` should be doable as simple index lookups when doing per memory bucket. Just have to translate the old bucket to the new ones (should normally be only 2, but I guess could be 4 or 8 even).

Reply:
Yeah, I expect resizing is really rare so I'm not really bothered about the cost of iteration when interrupted by a resize.

I didn't mention the cleaner code, but it was on my mind as well when I made the suggestion. I'm not so sure now after seeing the perf difference is > 5%, but I'm still leaning towards locking per bucket because it's one less knob to worry about.

Reply:
Yeah, I have already changed it over to per bucket.

Reply:
I just have to rewrite the resize iteration code to be more efficient. But I think I will be pretty happy with it.

Reply:
Okay, I have a new version that only reads the specific buckets even during a resize, which seems to pass the tests I have for now :)